### ISSUE TITLE
How to specify metric to optimize for without using (deprecated) TrainResult?

### ISSUE THREAD

**Issue by MarioIshac** [PyTorchLightning/pytorch-lightning#5341]
## ❓ Questions and Help

In my `training_step`, I previously returned this `TrainResult`:

```
train_result = TrainResult(train_log_dict["mae"])
train_result.log_dict(train_log_dict, on_epoch=True)

return train_result
```

But in migrating from 0.9.0 to 1.1.2, I learned that `TrainResult` was deprecated in favor of invoking `log_dict` on `self` like so:

```
self.log_dict(train_log_dict, on_epoch=True)

return train_log_dict
```
But now I'm no longer able to specify which metric I want training to be optimizing over. I ran into the error described in [#3113](https://github.com/PyTorchLightning/pytorch-lightning/issues/3113) as a result of this, but the solution provided is for pre-1.0.0.

How can I accomplish what I was previously doing? Apologies if this part of the migration was mentioned in the changelog, couldn't find anything describing how to replicate this particular aspect of `TrainResult`.

**Comment by awaelchli** [comment]
Return a single tensor or a dict with the keywoard "loss".

**Comment by MarioIshac** [comment]
> Return a single tensor or a dict with the keyword "loss".

Been a long time since I worked with my codebase, so might be misremembering here, but was there a change in 1.0 that forced the optimized-over metric to be named "loss" (compared to something like "mae" in my example)? Or has it been like this all along?

**Comment by awaelchli** [comment]
Yes but is this really a big problem? Optimizers in pytorch do minimization, hence we call it a loss. If you need to maximize, then pass the negative value. 

```python
def training_step(self, batch, batch_idx):
    x, y = batch
    out = self(x)
    mae = objective_function(out, y)
    return mae
    # or if you want to maximize mae
    return -mae
    # or if you want a dict
    return {"loss": mae}
    # or if you want to maximize mae
    return {"loss": -mae}
```

**Comment by MarioIshac** [comment]
> Yes but is this really a big problem? Optimizers in pytorch do minimization, hence we call it a loss. If you need to maximize, then pass the negative value.
> 
> ```python
> ...
> ```

I meant in regards to the naming instead of the direction. For my actual usecase, I have a parameter that decides whether I want to optimize over MAE or MSE. Pre-1.0, I would have two metrics: "mae" and "mse", and this parameter would decide whether `TrainResult` holds "mae" or "mse" as the to-be-optimized-over metric. Didn't need a devoted "loss" metric.

So to clarify: to make my program compatible with the new version I should add a third metric now called "loss"? And copy "mae"s value or "mse"s value depending on that parameter's value? No problem with that, just want to make sure that I'm not missing something.

**Comment by awaelchli** [comment]
> So to clarify: to make my program compatible with the new version I should add a third metric now called "loss"? And copy "mae"s value or "mse"s value depending on that parameter's value? No problem with that, just want to make sure that I'm not missing something.

yes, that's essentially what you have to do :) you can conditionally either return mae or mse. 
what used to be the value of "minimize" argument in TrainResult is now simply the return value in the training_step method. 

feel free to post some code snippet if you want me to look at your specific use case :)

**Comment by MarioIshac** [comment]
Alright that clears up everything :) Thanks for the help.
